### PR TITLE
bpo-40607: Reraise exception during task cancelation in asyncio.wait_for()

### PR DESCRIPTION
**Pull request by romasku**
https://bugs.python.org/issue40607

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@romasku

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Lib/asyncio/tasks.py [hunk 1]
```diff
@@ -496,6 +496,11 @@ async def wait_for(fut, timeout, *, loop=None):
             # after wait_for() returns.
             # See https://bugs.python.org/issue32751
             await _cancel_and_wait(fut, loop=loop)
+            # In case task cancellation failed with some
+            # exception, we should re-raise it
+            # See https://bugs.python.org/issue40607
+            if not fut.cancelled() and fut.exception() is not None:
+                raise fut.exception()
```
Comment (belkka) [inline Lib/asyncio/tasks.py]: Could it be
```python3
if not fut.cancelled():
    return fut.result()
```
?

File: Lib/asyncio/tasks.py [hunk 2]
```diff
@@ -496,6 +496,11 @@ async def wait_for(fut, timeout, *, loop=None):
             # after wait_for() returns.
             # See https://bugs.python.org/issue32751
             await _cancel_and_wait(fut, loop=loop)
+            # In case task cancellation failed with some
+            # exception, we should re-raise it
```
Comment (belkka) [inline Lib/asyncio/tasks.py:500]: What about the case when task cancellation failed with some **result** (not exception)? Is it OK to ignore it? I mean something like
```python3
async def inner():
    try:
        await asyncio.sleep(0.2)
    except asyncio.CancellationError:
        return 123  # instead of `raise FooException`
```

File: Lib/asyncio/tasks.py (hunk 1 above)
Comment (cjerdonek) [reply Lib/asyncio/tasks.py]: Building on this, how about--

```python
try:
    return fut.result()
except CancelledError as exc:
    raise TimeoutError() from exc
```

What's nice about this is that you'll see the `CancelledError` that caused the `TimeoutError` in the traceback. This will be even better when this PR is merged: https://github.com/python/cpython/pull/19951
In the traceback, you'll get to see exactly what line of code was interrupted by the cancellation.

Comment (romasku) [reply Lib/asyncio/tasks.py]: OK, thanks, I find this even more appropriate way to handle such situations.

**Comment by romasku** [comment]
I signed CLA yesterday, but for some reason [checker](https://check-python-cla.herokuapp.com/) returns 500 for my username? Is it OK and I should wait some more time?

**Comment by cjerdonek** [comment]
@Mariatta, are you able to answer @romasku's question above about the checker returning 500? Thanks a lot if you're able to help.

**Review by cjerdonek — CHANGES_REQUESTED** [review]
A couple higher-level comments before a more detailed review:

1. In the `TimeoutError` case, it would be good to check that the `CancelledError` is getting chained. You can see an example of how to check for that in a test here: https://github.com/python/cpython/pull/19858
Maybe this can be added to another existing test. I'm not sure as I haven't looked.

2. It looks like your tests take 0.1 seconds to run, which adds up a lot given the number of tests. It would be better to design them so they run without having to sleep for non-negligible amounts of time but still be tests that are (1) simple and (2) not flaky. This may take a little more creativity, or it might be easy. I'm not sure, but it should be doable.

**Comment by Mariatta** [comment]
Seems to work fine when I entered `romasku` in the CLA checker app.
It says that there is no b.p.o user associated with the has `romasku` GitHub user. Basically you need to add `romasku` to your b.p.o account.

File: Lib/asyncio/tasks.py [hunk 3]
```diff
@@ -496,7 +496,13 @@ async def wait_for(fut, timeout, *, loop=None):
...
             # See https://bugs.python.org/issue32751
             await _cancel_and_wait(fut, loop=loop)
-            raise exceptions.TimeoutError()
+            # In case task cancellation produced some result
+            # (either value or exception), we should pass through it
+            # See https://bugs.python.org/issue40607
+            try:
+                return fut.result()
+            except exceptions.CancelledError as exc:
+                raise exceptions.TimeoutError() from exc
```
Comment (1st1) [inline Lib/asyncio/tasks.py:505]: I'd put also an `else: raise TimeoutError`

**Comment by romasku** [comment]
I entered "@romasku" (with '@' symbol) as my github username in b.p.o. account, and was checking against it. Now everything works, thank you!

File: Lib/asyncio/tasks.py (hunk 3 above)
Comment (romasku) [reply Lib/asyncio/tasks.py:505]: Sorry, but I think that `else` branch is unreachable: if there isn't any exception in try block, then `return` works normally and we leave this method. Please explain if I am wrong.

**Comment by romasku** [comment]
Thank you for your review!
To overcome the second issue (about sleeping in tests) I've added `PositiveZero` class: it passes greater than zero check, so `asyncio.wait_for` main case logic runs, but it makes sleep time effectively zero.
To be honest, I am not sure about this solution: just using really small sleep time (for `0.000001` seconds) works the same way.
Also, I noticed that there are many other tests in `test_tasks.py` that sleep for 0.1-0.3 seconds - maybe it's better to make another PR to optimize those tests?

**Comment by romasku** [comment]
I have made the requested changes; please review again

File: Lib/asyncio/tasks.py (hunk 3 above)
Comment (1st1) [reply Lib/asyncio/tasks.py:505]: Yes, but your new code is simply not equivalent to the previous code that raised TimeoutError unconditionally. Now I see that there might be a condition for clear return and I don't want that.

**Review by 1st1 — CHANGES_REQUESTED** [review]

File: Lib/asyncio/tasks.py (hunk 3 above)
Comment (1st1) [reply Lib/asyncio/tasks.py:505]: The linked issue talks about the correct exception propagation. Basically, `wait_for` timeouts with an exception and we want that original exception to be propagated. That's fine and I'm +1 for that.

The code here is slightly different though and a bit hard to trace so it would be easier to read if it was replaced with:

```python
           try:
                fut.result()
            except exceptions.CancelledError as exc:
                raise exceptions.TimeoutError() from exc
            else:
                raise exceptions.TimeoutError()
```

Comment (cjerdonek) [reply Lib/asyncio/tasks.py:505]: I think the issue here is that there is a race condition where the task can complete after the timeout has elapsed. So the question is whether the caller should get to learn that the task completed in this case.

Comment (1st1) [reply Lib/asyncio/tasks.py:505]: Yes, the current implementation discards the result, as the probability of actually having it is low. I'd like to keep it that way. If we want to change the current behavior that should be done in a separate issue, with a new PR and new discussion.

**Comment by romasku** [comment]
I have made the requested changes; please review again

File: Lib/test/test_asyncio/test_tasks.py [hunk 4]
```diff
@@ -80,6 +80,22 @@ def __await__(self):
         return self
 
 
+@functools.total_ordering
+class PositiveZero:
```
Comment (cjerdonek) [inline Lib/test/test_asyncio/test_tasks.py]: This is a clever idea. But like you hinted at earlier, it's probably overkill and you could just use a small constant instead. How about defining something like `_EPSILON = 0.0001` at the top of the file, with a similar comment as below for what it's used for. A constant like this should probably only be used in cases where test flakiness can't arise after analyzing the test. Otherwise, the small number needs to be chosen on a case-by-case basis. (I haven't done the analysis for your test to know if flakiness is possible.)

By the way, after I wrote my suggestion about using a smaller timeout, like you said, I realized that _some_ hard-coded number had to be picked because it's being used as an argument. So my suggestion wasn't really valid for your case (but it can be used in other cases).

File: Lib/test/test_asyncio/test_tasks.py [hunk 5]
```diff
@@ -877,10 +893,51 @@ async def inner():
 
             inner_task = self.new_task(loop, inner())
 
-            with self.assertRaises(asyncio.TimeoutError):
-                await asyncio.wait_for(inner_task, timeout=0.1)
+            with self.assertRaises(asyncio.TimeoutError) as cm:
```
Comment (cjerdonek) [inline Lib/test/test_asyncio/test_tasks.py]: Can these assertions be moved outside of the coroutine and after the call to `loop.run_until_complete()`? One advantage of doing it that way is that it's easier to see that your assertions will always be exercised, as opposed to skipping them due to e.g. an early return. It also makes the coroutine smaller. This comment holds for the other test cases as well.

File: Misc/NEWS.d/next/Library/2020-05-13-15-32-13.bpo-40607.uSPFCi.rst [hunk 6]
```diff
@@ -0,0 +1,2 @@
+When cancelling the task due to a timeout, :meth:`asyncio.wait_for` will now
+reraise exceptions triggered by custom cancellation handling.
```
Comment (cjerdonek) [inline Misc/NEWS.d/next/Library/2020-05-13-15-32-13.bpo-40607.uSPFCi.rst]: How about this wording?

> When cancelling a task due to timeout, :meth:`asyncio.wait_for` will now raise the exception if an error happens during cancellation.

I think it would also be good to update the `wait_for()` docs with the new behavior.

**Review by cjerdonek — CHANGES_REQUESTED** [review]
Some final comments.

File: Misc/NEWS.d/next/Library/2020-05-13-15-32-13.bpo-40607.uSPFCi.rst (hunk 6 above)
Comment (romasku) [reply Misc/NEWS.d/next/Library/2020-05-13-15-32-13.bpo-40607.uSPFCi.rst]: I am not sure about your rewording:
- `asyncio.wait_for` already raises an exception (`TimeoutError`) if an error happens during cancellations
- it can be misunderstood as it will raise some "special" exception, something like `CancellationFailedException`.

My idea is to outline that same exception is propagated. I think the following version is more accurate:
> When cancelling a task due to timeout, :meth:asyncio.wait_for will now **propagate** the exception if an error happens during cancellation.

**Comment by romasku** [comment]
I have made the requested changes; please review again

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
This looks great now. @cjerdonek please review again.

File: Lib/test/test_asyncio/test_tasks.py [hunk 7]
```diff
@@ -80,6 +80,12 @@ def __await__(self):
         return self
 
 
+# Next value can be used as very small sleep time:
+# it passes check "sleep_time > 0", but has almost
```
Comment (cjerdonek) [inline Lib/test/test_asyncio/test_tasks.py]: I think it would be better to say timeout or delay here instead of sleep time because it was the timeout inside `wait_for()` that has to be > 0, and it's being passed to `call_later()` rather than `sleep()`. (And often for 
 sleep, `sleep(0)` is sufficient so an epsilon isn't needed there.)

File: Lib/test/test_asyncio/test_tasks.py [hunk 8]
```diff
@@ -868,6 +874,8 @@ def test_wait_for_waits_for_task_cancellation(self):
         task_done = False
 
         async def foo():
+            nonlocal chained
```
Comment (cjerdonek) [inline Lib/test/test_asyncio/test_tasks.py]: Is this `nonlocal` still needed?

**Review by cjerdonek — CHANGES_REQUESTED** [review]
There are a couple more tiny things I noticed, but otherwise looks great. Thanks a lot for your good work on this, @romasku!

**Comment by romasku** [comment]
I have made the requested changes; please review again

**Review by cjerdonek — APPROVED** [review]
Thanks, @romasku.

**Comment by cjerdonek** [comment]
@romasku, I forgot there are a couple more administrative things. Can you add your name to `Misc/ACKS` and also add "Patch by <your name>." at the end of your NEWS entry? More details can be found here: https://devguide.python.org/committing/#handling-others-code

**Comment by romasku** [comment]
@cjerdonek, done. I am happy to contribute!

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
Thanks so much, Roman!
